File: app/src/main/java/cz/johnyapps/cheers/views/EmptyMessageRecyclerView.java
package cz.johnyapps.cheers.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.recyclerview.widget.RecyclerView;

import cz.johnyapps.cheers.R;
import cz.johnyapps.cheers.adapters.recycleradapters.BaseAdapter;

public class EmptyMessageRecyclerView extends LinearLayout {
    @Nullable
    private String emptyMessage = null;
    private RecyclerView recyclerView;
    private View emptyMessageView;

    public EmptyMessageRecyclerView(@NonNull Context context) {
        super(context);
        init(null, 0);
    }

    public EmptyMessageRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public EmptyMessageRecyclerView(@NonNull Context context, @Nullable AttributeSet attrs, int theme) {
        super(context, attrs, theme);
        init(attrs, theme);
    }

    public void init(@Nullable AttributeSet attrs, int theme) {
        LayoutInflater.from(getContext()).inflate(R.layout.view_empty_message_recycler_view,
                this,
                true);

        if (attrs != null) {
            final TypedArray array = getContext().obtainStyledAttributes(attrs,
                    R.styleable.EmptyMessageRecyclerView,
                    theme,
                    0);

            emptyMessage = array.getString(R.styleable.EmptyMessageRecyclerView_emptyMessage);
            array.recycle();
        }

        AppCompatTextView emptyMessageTextView = findViewById(R.id.emptyMessageTextView);
        assert emptyMessageTextView != null;
        emptyMessageTextView.setText(emptyMessage);

        recyclerView = findViewById(R.id.recyclerView);
        emptyMessageView = findViewById(R.id.emptyMessageLayout);
    }

    private void emptyMessageVisible(boolean visible) {
        if (visible) {
            recyclerView.setVisibility(GONE);
            emptyMessageView.setVisibility(VISIBLE);
        } else {
            recyclerView.setVisibility(VISIBLE);
            emptyMessageView.setVisibility(GONE);
        }
    }

    public void setAdapter(RecyclerView.Adapter<? extends RecyclerView.ViewHolder> adapter) {
        if (adapter instanceof BaseAdapter) {
            BaseAdapter baseAdapter = (BaseAdapter) adapter;
            baseAdapter.setOnDataSetChangedListener(this::emptyMessageVisible);
        }

        recyclerView.setAdapter(adapter);
    }

    public void setLayoutManager(RecyclerView.LayoutManager layoutManager) {
        recyclerView.setLayoutManager(layoutManager);
    }

    public void scrollToPosition(int position) {
        recyclerView.scrollToPosition(position);
    }

    @Override
    public void setOnScrollChangeListener(OnScrollChangeListener l) {
        recyclerView.setOnScrollChangeListener(l);
    }
}

File: app/src/main/java/cz/johnyapps/cheers/views/EmptyMessageRecyclerView.kt
package cz.johnyapps.cheers.views

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import cz.johnyapps.cheers.R
import cz.johnyapps.cheers.adapters.BaseAdapter
import cz.johnyapps.cheers.databinding.ViewEmptyMessageRecyclerViewBinding

class EmptyMessageRecyclerView: LinearLayout {
    private lateinit var binding: ViewEmptyMessageRecyclerViewBinding

    constructor(context: Context) : super(context) {
        initialize(null, 0)
    }

    constructor(context: Context, attrs: AttributeSet?): super(context, attrs) {
        initialize(attrs, 0)
    }

    constructor(context: Context, attrs: AttributeSet?, theme: Int): super(context, attrs, theme) {
        initialize(attrs, theme)
    }

    private fun initialize(attrs: AttributeSet?, theme: Int) {
        var emptyMessage: String? = null

        if (attrs != null) {
            val array = context.obtainStyledAttributes(attrs, R.styleable.EmptyMessageRecyclerView, theme, 0)

            emptyMessage = array.getString(R.styleable.EmptyMessageRecyclerView_emptyMessage)

            array.recycle()
        }

        binding = DataBindingUtil.inflate(LayoutInflater.from(context),
            R.layout.view_empty_message_recycler_view,
            this,
            true)
        binding.emptyMessageTextView.text = emptyMessage
    }

    fun setAdapter(adapter: RecyclerView.Adapter<*>, layoutManager: RecyclerView.LayoutManager) {
        if (adapter is BaseAdapter<*, *>) {
            adapter.onListChangeListener = object : BaseAdapter.OnListChangedListener {
                override fun onChange(isListEmpty: Boolean) {
                    if (isListEmpty) {
                        binding.recyclerView.visibility = View.GONE
                        binding.emptyMessageLayout.visibility = View.VISIBLE
                    } else {
                        binding.recyclerView.visibility = View.VISIBLE
                        binding.emptyMessageLayout.visibility = View.GONE
                    }
                }
            }
        }

        binding.recyclerView.layoutManager = layoutManager
        binding.recyclerView.adapter = adapter
    }

    fun scrollToPosition(position: Int) {
        binding.recyclerView.scrollToPosition(position)
    }
}